File: android/src/main/kotlin/dev/sonerik/camera2/CropUtils.kt
package dev.sonerik.camera2

import android.graphics.Rect

fun centerCroppedSourceRect(
        sourceWidth: Float,
        sourceHeight: Float,
        targetWidth: Float,
        targetHeight: Float,
        inSourceCoordinates: Boolean = true
): Rect {
    val sourceAspectRatio = sourceWidth / sourceHeight
    val targetAspectRatio = targetWidth / targetHeight
    val widthFactor = sourceWidth / targetWidth
    val heightFactor = sourceHeight / targetHeight
    val scale = if (sourceAspectRatio <= targetAspectRatio) widthFactor else heightFactor
    val targetSourceWidth = sourceWidth / scale
    val targetSourceHeight = sourceHeight / scale
    val extraTargetWidth = targetSourceWidth - targetWidth
    val extraSourceWidth = extraTargetWidth * scale
    val extraTargetHeight = targetSourceHeight - targetHeight
    val extraSourceHeight = extraTargetHeight * scale

    if (inSourceCoordinates) {
        val width = sourceWidth - extraSourceWidth
        val height = sourceHeight - extraSourceHeight
        val left = extraSourceWidth / 2
        val top = extraSourceHeight / 2
        val right = left + width
        val bottom = top + height
        return Rect(left.toInt(), top.toInt(), right.toInt(), bottom.toInt())
    } else {
        val width = targetWidth - extraTargetWidth
        val height = targetHeight - extraTargetHeight
        val left = extraTargetWidth / 2
        val top = extraTargetHeight / 2
        val right = left + width
        val bottom = top + height
        return Rect(left.toInt(), top.toInt(), right.toInt(), bottom.toInt())
    }
}

fun centerCroppedStencilRect(
        rect: Rect,
        stencilWidthPercent: Float,
        stencilAspectRatio: Float
): Rect {
    val targetCropWidth = rect.width() * stencilWidthPercent
    val targetCropHeight = targetCropWidth / stencilAspectRatio
    val targetCropExtraWidth = rect.width() - targetCropWidth
    val targetCropExtraHeight = rect.height() - targetCropHeight

    val left = rect.left + targetCropExtraWidth / 2
    val top = rect.top + targetCropExtraHeight / 2
    val right = left + targetCropWidth
    val bottom = top + targetCropHeight
    return Rect(left.toInt(), top.toInt(), right.toInt(), bottom.toInt())
}

File: ios/Classes/CropUtils.swift
import Foundation
import AVKit

func centerCroppedSourceRect(sourceWidth: CGFloat, sourceHeight: CGFloat, targetWidth: CGFloat, targetHeight: CGFloat, inSourceCoordinates: Bool = true) -> CGRect {
    let sourceAspectRatio = sourceWidth / sourceHeight
    let targetAspectRatio = targetWidth / targetHeight
    let widthFactor = sourceWidth / targetWidth
    let heightFactor = sourceHeight / targetHeight
    let scale = sourceAspectRatio <= targetAspectRatio ? widthFactor : heightFactor
    let targetSourceWidth = sourceWidth / scale
    let targetSourceHeight = sourceHeight / scale
    let extraTargetWidth = targetSourceWidth - targetWidth
    let extraSourceWidth = extraTargetWidth * scale
    let extraTargetHeight = targetSourceHeight - targetHeight
    let extraSourceHeight = extraTargetHeight * scale
    
    if (inSourceCoordinates) {
        return CGRect(
            x: extraSourceWidth / 2,
            y: extraSourceHeight / 2,
            width: sourceWidth - extraSourceWidth,
            height: sourceHeight - extraSourceHeight
        )
    } else {
        return CGRect(
            x: extraTargetWidth / 2,
            y: extraTargetHeight / 2,
            width: targetWidth - extraTargetWidth,
            height: targetHeight - extraTargetHeight
        )
    }
}

func centerCroppedStencilRect(rect: CGRect, stencilWidthPercent: CGFloat, stencilAspectRatio: CGFloat) -> CGRect {
    let targetCropWidth = rect.width * stencilWidthPercent
    let targetCropHeight = targetCropWidth / stencilAspectRatio
    let targetCropExtraWidth = rect.width - targetCropWidth
    let targetCropExtraHeight = rect.height - targetCropHeight
    return CGRect(
        x: rect.minX + targetCropExtraWidth / 2,
        y: rect.minY + targetCropExtraHeight / 2,
        width: targetCropWidth,
        height: targetCropHeight
    )
}
